internal/graph/check.go

```diff
@@ -848,6 +848,233 @@ func (c *LocalChecker) checkComputedUserset(_ context.Context, req *ResolveCheck
+
+		tupleKey := &openfgav1.TupleKey{
+			Object:   userObj,
+			Relation: computedRelation,
+			User:     tk.GetUser(),
+		}
+
+		if _, err := typesys.GetRelation(tuple.GetType(userObj), computedRelation); err != nil {
+			if errors.Is(err, typesystem.ErrRelationUndefined) {
+				continue // skip computed relations on tupleset relationships if they are undefined
+			}
+		}
+
+		// Note: we add TTU read below
+		handlers = append(handlers, c.dispatch(ctx, req, tupleKey))
+	}
+
+	if len(handlers) == 0 && errs != nil {
+		telemetry.TraceError(span, errs)
+		return nil, errs
+	}
+
+	resp, err := union(ctx, c.concurrencyLimit, handlers...)
+	if err != nil {
+		telemetry.TraceError(span, err)
+		return nil, errors.Join(errs, err)
+	}
+
+	// if we had 3 dispatched requests, and the final result is "allowed = false",
+	// we want final reads to be (N1 + N2 + N3 + 1) and not (N1 + 1) + (N2 + 1) + (N3 + 1)
+	// if final result is "allowed = true", we want final reads to be N1 + 1
+	resp.GetResolutionMetadata().DatastoreQueryCount++
+
+	return resp, nil
+}
+
+// checkTTUFastPath is the fast path for checkTTU where we can short-circuit TTU evaluation.
+// This requires both the TTU's tuplesetRelation and computedRelation be exclusively directly assignable.
+// The general idea is to check whether user has relation with the specified TTU by finding object
+// intersection between tuplesetRelation's object AND objectType's computedRelation for user.  For example,
+//
+//	type group
+//	  define member: [user]
+//	type doc
+//	  define parent: [group]
+//	  define viewer: member from parent
+//
+// check(user, viewer, doc) will find the intersection of all group assigned to the doc's parent AND
+// all group where the user is a member of.
+func (c *LocalChecker) checkTTUFastPath(ctx context.Context, req *ResolveCheckRequest, rewrite *openfgav1.Userset, iter storage.TupleKeyIterator) (*ResolveCheckResponse, error) {
+	var errs error
+
+	typesys, ok := typesystem.TypesystemFromContext(ctx)
+	if !ok {
+		return nil, fmt.Errorf("typesystem missing in context")
+	}
+
+	ds, ok := storage.RelationshipTupleReaderFromContext(ctx)
+	if !ok {
+		return nil, fmt.Errorf("relationship tuple reader datastore missing in context")
+	}
+
+	computedRelation := rewrite.GetTupleToUserset().GetComputedUserset().GetRelation()
+	tk := req.GetTupleKey()
+
+	response := &ResolveCheckResponse{
+		Allowed: false,
+		ResolutionMetadata: &ResolveCheckResponseMetadata{
+			DatastoreQueryCount: req.GetRequestMetadata().DatastoreQueryCount + 1,
+		},
+	}
+
+	// tuplesetRelationUserMap contains of users that fits the tuplesetRelation.  For example,
+	// type group
+	//   define member: [user]
+	// type doc
+	//   define parent: [group]
+	//   define viewer: member from parent
+	// tuplesetRelationUserMap will list all group for doc:1's parent
+	// [group][1]
+	// [group][2]
+	// [group][3]
+	tuplesetRelationUserMap := make(map[string]map[string]struct{})
```

Comment:
nit: this example doesn't make a lot of sense, maybe change the model so that `define parent: [group1, group2]` and then it's clear the map is like

````
group1 -> 1,2,3
group2 -> 4,5
```

Reply:
The reasoning is to use a `Map[Set]` so that lookups in the ID are faster